build_modules/lib/src/module_builder.dart

```diff
@@ -7,15 +7,95 @@ import 'dart:convert';
+final _readMetas = new Resource<Set<AssetId>>(() => new Set<AssetId>(),
+    dispose: (s) => s.clear());
+
+/// Map of [AssetId] to corresponding primary [AssetId] within the same
+/// [Module].
+final _assetToPrimary = new Resource<Map<AssetId, AssetId>>(() => {},
+    dispose: (map) => map.clear());
+
+void _cacheAssetToPrimary(
+    MetaModule meta, Map<AssetId, AssetId> assetToPrimary) {
+  for (var module in meta.modules) {
+    for (var source in module.sources) {
+      assetToPrimary[source] = module.primarySource;
+    }
+  }
+}
+
+/// Removes dependencies from the provided [Module] so that only those with
+/// corresponding module files remain.
+Future<Module> _cleanModuleDeps(
+    BuildStep buildStep, Module module, Map<AssetId, AssetId> assetToPrimary,
+    {bool isCourseStrategy: false}) async {
+  var cleanedDeps = new Set<AssetId>();
+  for (var dep in module.directDependencies) {
+    if (!isCourseStrategy) {
+      // Since we are not using the ccourse strategy we can safely add
+      // all dependencies in the same package as they will have a
+      // corresponding module file.
+      if (dep.package == module.primarySource.package) {
+        cleanedDeps.add(dep);
+        continue;
+        // It is possible that this dep came from another fine grained
+        // module. Look for the corresponding module file.
+      } else if (await buildStep
+          .canRead(dep.changeExtension(moduleExtension))) {
+        cleanedDeps.add(dep);
+        continue;
+      }
+    }
+    var primaryDep = assetToPrimary[dep];
+    if (primaryDep == null) {
+      // We don't know the primary source for the dep so look for it
+      // using the meta module file.
+      var depMeta = new MetaModule.fromJson(json.decode(
+              await buildStep.readAsString(
+                  new AssetId(dep.package, 'lib/$metaModuleExtension')))
+          as Map<String, dynamic>);
+      _cacheAssetToPrimary(depMeta, assetToPrimary);
+    }
+    cleanedDeps.add(assetToPrimary[dep]);
+  }
+  return new Module(module.primarySource, module.sources, cleanedDeps);
+}
+
+/// Processes a [MetaModule] and adds to the [primaryToClean] cache.
+Future<Null> _cacheCleanedModules(
+    BuildStep buildStep,
+    MetaModule meta,
+    Map<AssetId, Module> primaryToClean,
+    Map<AssetId, AssetId> assetToPrimary) async {
+  _cacheAssetToPrimary(meta, assetToPrimary);
+  for (var module in meta.modules) {
+    primaryToClean[module.primarySource] = await _cleanModuleDeps(
+        buildStep, module, assetToPrimary,
+        isCourseStrategy: true);
+  }
+}
+
 /// Creates `.module` files for any `.dart` file that is the primary dart
 /// source of a [Module].
 class ModuleBuilder implements Builder {
-  const ModuleBuilder();
+  final bool _isCoarse;
+  const ModuleBuilder({bool isCourse: false}) : _isCoarse = isCourse;
+
+  factory ModuleBuilder.forOptions(BuilderOptions options) {
+    return new ModuleBuilder(isCourse: options.config['strategy'] == 'coarse');
```

Comment:
Lets add some validation around the acceptable values for `strategy`, otherwise typos can be frustrating.

Also, lets validate that we don't see any unrecognized top level keys.